src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp

```diff
@@ -40,10 +41,13 @@ WriteBufferFromAzureBlobStorage::~WriteBufferFromAzureBlobStorage()
     finalize();
 }
 
-void WriteBufferFromAzureBlobStorage::execWithRetry(std::function<void()> func, size_t num_tries)
+void WriteBufferFromAzureBlobStorage::execWithRetry(std::function<void()> func, size_t num_tries, size_t cost)
 {
-    auto handle_exception = [&](const auto & e, size_t i)
+    auto handle_exception = [&, this](const auto & e, size_t i)
     {
+        if (cost)
+            write_settings.resource_link.accumulate(cost); // Accumulate resource for later use, because we have failed to consume it
```

Comment:
probably we should also call `resource_link.accumulate` below when we catch exceptions thrown by `func()` and decide ti re throw it in case last iteration came
```
if (i == num_tries - 1)
             throw;
```

Reply:
You mean any other exception beside those two already handled? Okay let's do it for consistency. Actually, it is hard to tell if resource was consumed in case of a failure, it depends on what happened.